File: fabric-audit-agent/core/correlate.js
/**
 * Group related active findings into root-cause clusters with a narrative. Pure.
 * @param {object[]} findings  active findings (with .key)
 * @returns {Array<{theme:string, findingKeys:string[], narrative:string}>}
 */
export function correlate(findings = []) {
  const has = (prefix) => findings.filter(f => typeof f.key === 'string' && f.key.startsWith(prefix));
  const correlations = [];

  // 1. Capacity pressure: a throttle driven by contention / oversized models.
  const throttle = has('capacity.throttle');
  const drivers = [...has('capacity.contention'), ...has('capacity.oversized-model')];
  if (throttle.length && drivers.length) {
    correlations.push({
      theme: 'capacity-pressure',
      findingKeys: [...throttle, ...drivers].map(f => f.key),
      narrative: `Capacity throttling is likely driven by ${drivers.length} optimization issue(s) — resolve those before sizing up the SKU.`,
    });
  }

  // 2. Refresh chain: failures spanning models AND pipelines = shared upstream.
  const modelFail = has('model.refresh-failing');
  const pipeFail = has('pipeline.failing');
  if (modelFail.length && pipeFail.length) {
    correlations.push({
      theme: 'refresh-chain',
      findingKeys: [...modelFail, ...pipeFail].map(f => f.key),
      narrative: `Refresh failures span ${modelFail.length} model(s) and ${pipeFail.length} pipeline(s) — likely a shared gateway/source. Investigate the upstream together.`,
    });
  }

  // 3. Security cluster: multiple access/security findings = one access review.
  const sec = has('security.');
  if (sec.length >= 2) {
    correlations.push({
      theme: 'security-cluster',
      findingKeys: sec.map(f => f.key),
      narrative: `${sec.length} security/access findings detected together — handle as one access-review action.`,
    });
  }

  return correlations;
}

File: fabric-audit-agent-app/fabric_audit_agent/correlate.py
"""Group related findings into root-cause clusters. Port of ``core/correlate.js``. Pure."""


def correlate(findings=None):
    findings = findings or []

    def has(prefix):
        return [f for f in findings if isinstance(f.get("key"), str) and f["key"].startswith(prefix)]

    correlations = []

    throttle = has("capacity.throttle")
    drivers = has("capacity.contention") + has("capacity.oversized-model")
    if throttle and drivers:
        correlations.append({
            "theme": "capacity-pressure",
            "findingKeys": [f["key"] for f in (throttle + drivers)],
            "narrative": f"Capacity throttling is likely driven by {len(drivers)} optimization issue(s) — resolve those before sizing up the SKU.",
        })

    model_fail = has("model.refresh-failing")
    pipe_fail = has("pipeline.failing")
    if model_fail and pipe_fail:
        correlations.append({
            "theme": "refresh-chain",
            "findingKeys": [f["key"] for f in (model_fail + pipe_fail)],
            "narrative": f"Refresh failures span {len(model_fail)} model(s) and {len(pipe_fail)} pipeline(s) — likely a shared gateway/source. Investigate the upstream together.",
        })

    sec = has("security.")
    if len(sec) >= 2:
        correlations.append({
            "theme": "security-cluster",
            "findingKeys": [f["key"] for f in sec],
            "narrative": f"{len(sec)} security/access findings detected together — handle as one access-review action.",
        })

    return correlations
